File: Clasificare/Category.py
from pydoc import describe
from nltk import *
import pickle
from keras import models, preprocessing as kprocessing
import json
import nltk

from confluent_kafka import Consumer
from confluent_kafka import Producer

nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')

STOPWORDS_CORPUS = corpus.stopwords.words('english')
LEMMATIZATION    = WordNetLemmatizer()
STEMMING         = PorterStemmer()

with open('tokenizer.pickle', 'rb') as handle:
    tokenizer = pickle.load(handle)

model = models.load_model('model2.h5')

labels=['ARTS & CULTURE', 'BUSINESS', 'CRIME', 'EDUCATION',
       'ENTERTAINMENT', 'ENVIRONMENT', 'FOOD & DRINK', 'HEALTHY LIVING',
       'HOME & LIVING', 'POLITICS', 'RELIGION', 'SCIENCE', 'SPORTS',
       'STYLE & BEAUTY', 'TRAVEL']


##############################
conf = {'bootstrap.servers': "10.0.0.1:9092",
        'group.id': "group1",
        'auto.offset.reset': 'earliest'}

consumer = Consumer(conf)
consumer.subscribe(["crawled_news"])

conf_p = {'bootstrap.servers': '10.0.0.1:9092'}
producer = Producer(conf_p)

##############################

def preprocess_string(text, lemmatizer = WordNetLemmatizer(), stemmer = PorterStemmer(), stopwords_corpus = corpus.stopwords.words('english')):
    text = str(text).lower().strip()
    text = re.sub(r'[^\w\s]', '', text).split()
    text = [word for word in text if word not in stopwords_corpus]
    text = [lemmatizer.lemmatize(word) for word in text]
    text = [stemmer.stem(word) for word in text]
    return text

while True:
    msg = consumer.poll(timeout=1.0)
    if msg is None: continue
    else:
        try:
            news_json = json.loads(msg.value())

            paragrafe=news_json["content"].split("*NewPARAGRAF*")

            title=news_json["title"]
            description=news_json["description"]

            paragrafe.insert(0,title)
            paragrafe.insert(0,description)

            paragrafe=list(map(preprocess_string, paragrafe))

            while("" in paragrafe) :
                 paragrafe.remove("")
                 
            tokParagrafe=tokenizer.texts_to_sequences(paragrafe)
	    
            padParagrafe=kprocessing.sequence.pad_sequences(tokParagrafe,  maxlen=128, padding="post", truncating="post")
            
            results=model.predict(padParagrafe)

            vector=[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
            nr_paragrafe=0
            for i in range(len(results)):
                nr_paragrafe+=1*len(tokParagrafe[i])
                results[i]=results[i]*len(tokParagrafe[i])
                for j in range(len(vector)):
                    vector[j]+=results[i][j]
                
            for i in range(len(vector)):
                vector[i]=vector[i]/nr_paragrafe
            
    
            count=0
            for i in range(len(vector)):
                if(vector[i]>=0.33):
                    count+=1

            Category=""    
            for i in range(len(vector)):
                if(vector[i]>=0.25):
                    Category=Category+labels[i]+"*"+str(vector[i])+"-"
            
            if(Category==""):
                Category="No category"
                
            print(Category)
            
            news_json["category"]=Category

            producer.poll(0)
            producer.produce('categorised_news', json.dumps(news_json))
        except Exception as e:
            print(e)
            print("Invalid")

File: Site_Web/interfata-web/src/MainPage/ViewNewsPage/Category.js

import React from "react";
import "./Category.css";

class Category extends React.Component {
  constructor(props) {
    super(props);
  }




  render() {

    let color = "white";

    if (this.props.Category === "SCIENCE") {
      color = "rgba(0,250,250,0.75)";

    }
    else if (this.props.Category === "ARTS & CULTURE") {
      color = "rgba(100,200,150,0.75)";
    }
    else if (this.props.Category === "BUSINESS") {
      color = "rgba(200,40,20,0.75)";
    }
    else if (this.props.Category === "TRAVEL") {
      color = "rgba(40,200,0,0.75)";
    }
    else if (this.props.Category === "POLITICS") {
      color = "rgba(255,50,50,0.75)";
    }
    else if (this.props.Category === "HEALTHY LIVING") {
      color = "rgba(200,250,250,0.75)";
    }
    else if (this.props.Category === "FOOD & DRINK") {
      color = "rgba(250,150,100,0.75)";
    }
    else if (this.props.Category === "CRIME") {
      color = "rgba(255,0,30,0.75)";
    }
    else if (this.props.Category === "HOME & LIVING") {
      color = "rgba(255,0,100,0.75)";
    }
    else if (this.props.Category === "RELIGION") {
      color = "rgba(155,75,75,0.75)";
    }
    else if (this.props.Category === "STYLE & BEAUTY") {
      color = "rgba(155,255,105,0.75)";
    }
    else if (this.props.Category === "SPORTS") {
      color = "rgba(30,100,255,0.75)";
    }
    else if (this.props.Category === "ENVIRONMENT") {
      color = "rgba(10,255,25,0.75)";
    }
    else if (this.props.Category === "ENTERTAINMENT") {
      color = "rgba(250,250,55,0.75)";
    }
    else if (this.props.Category === "EDUCATION") {
      color = "rgba(250,0,255,0.75)";
    }

    if (this.props.Size == "small") {
      return (
        <div className="Category-Outer" style={{ background: color, width: "90px", }}>
          <div className="Category-Category" style={{ fontSize: "10px" }}>
            {this.props.Category}
          </div>
          <div className="Category-Procent" style={{ fontSize: "10px" }}>
            {parseInt(this.props.Procent * 100)}%
          </div>
        </div>
      )
    }
    else {
      return (
        <div className="Category-Outer" style={{ background: color, width: "140px", }}>
          <div className="Category-Category" style={{ fontSize: "16px" }}>
            {this.props.Category}
          </div>
          <div className="Category-Procent" style={{ fontSize: "16px" }}>
            {parseInt(this.props.Procent * 100)}%
          </div>
        </div>
      )
    }
  }
}

export default Category;
